File: Integrations/sdks/python/scriptbee_analysis_sdk/abstractions/install_plugin_use_case.py
from abc import ABC, abstractmethod
from dataclasses import dataclass

from scriptbee_analysis_sdk.abstractions.errors.invalid_plugin_error import InvalidPluginError
from scriptbee_analysis_sdk.abstractions.errors.plugin_installation_error import (
    PluginInstallationError,
)
from scriptbee_analysis_sdk.domain.plugins import PluginId


@dataclass(frozen=True)
class Success:
    pass


InstallPluginResult = Success | InvalidPluginError | PluginInstallationError


class InstallPluginUseCase(ABC):
    @abstractmethod
    def install_plugin(self, plugin_id: PluginId) -> InstallPluginResult: ...

File: Integrations/sdks/typescript/src/abstractions/install-plugin-use-case.ts
import type { PluginId } from "../domain/plugins.js";
import type { InvalidPluginError, PluginInstallationError } from "./errors.js";

export interface Success {
  readonly _tag: "Success";
}

export function success(): Success {
  return { _tag: "Success" };
}

export type InstallPluginResult = Success | InvalidPluginError | PluginInstallationError;

export interface InstallPluginUseCase {
  installPlugin(pluginId: PluginId): InstallPluginResult;
}
